Mainnet/DividendToken/DividendToken/DividendToken.cs

```diff
@@ -0,0 +1,288 @@
+﻿using Stratis.SmartContracts;
+using Stratis.SmartContracts.Standards;
+
+[Deploy]
+public class DividendToken : SmartContract, IStandardToken
+{
+
+    public ulong Dividends
+    {
+        get => PersistentState.GetUInt64(nameof(this.Dividends));
+        private set => PersistentState.SetUInt64(nameof(this.Dividends), value);
+    }
+
+    private Account GetAccount(Address address) => PersistentState.GetStruct<Account>($"Account:{address}");
+
+    private void SetAccount(Address address, Account account) => PersistentState.SetStruct($"Account:{address}", account);
+
+
+    public DividendToken(ISmartContractState state, ulong totalSupply, string name, string symbol)
+        : base(state)
+    {
+        this.TotalSupply = totalSupply;
+        this.Name = name;
+        this.Symbol = symbol;
+        this.SetBalance(Message.Sender, totalSupply);
+    }
+
+    /// <summary>
+    /// It is advised that deposit amount should to be evenly divided by total supply, 
+    /// otherwise small amount of satoshi may lost(burn)
+    /// </summary>
+    public override void Receive()
+    {
+        Dividends += Message.Value;
+    }
+
+    public bool TransferTo(Address to, ulong amount)
+    {
+        UpdateAccount(Message.Sender);
+        UpdateAccount(to);
+
+        return TransferTokensTo(to, amount);
+    }
+
+    public bool TransferFrom(Address from, Address to, ulong amount)
+    {
+        UpdateAccount(from);
+        UpdateAccount(to);
+
+        return TransferTokensFrom(from, to, amount);
+    }
+
+    Account UpdateAccount(Address address)
+    {
+        var account = GetAccount(address);
+        var newDividends = GetWithdrawableDividends(address, account);
+
+        if (newDividends > 0)
+        {
+            account.DividendBalance = checked(account.DividendBalance + newDividends);
+            account.CreditedDividends = Dividends;
```

Comment:
The CreditedDividends for this account should equal the entire value of all `Dividends`?

Reply:
I know it looks weird at first sight without knowing design decisions. I added description but it is not covering technical design decisions. 

Dividends incremented whenever contract get deposits by `DistributeDividends` or `Receive` methods.  But we don't calculate  address deposits in those methods because it requires iteration on all token holder addresses which is very costly and iterative code is bad design for smart contracts. The contract lazily compute dividends whenever tokens transferred By Transfer methods.  So each address tracks its own computed Dividends by `CreditedDividends`  If `Dividends == CreditedDividends`  then all deposited dividends already computed and added `DividendBalance`  or If it is not equal so there are some new dividends (`Dividends - CreditedDividends`) which is not yet computed. 


As you may see `DividendBalance`  should be divided by `TotalSupply` but it is not. Because it is saving divided value cause precision loss. So we save values without divide operation but we always display dividends by dividing to TotalSupply when user request it.